docs/rfcs/004-session-management.md

```diff
@@ -0,0 +1,125 @@
+
+> TODO: graphic
+
+### Add a command-line argument to explicitly choose the Session
+
+When opening Atom from the command line with one or more locations specified:
+
+1. If exactly one non-open Session matches those paths, that Session is loaded into the newly opened Atom window.
+2. If no non-open Sessions match, a new Session is created for that Atom window.
+3. If multiple Sessions match, a new Session is created for that Atom window, and a notification is displayed allowing the user to switch to any of the matching Sessions and destroy the just-created one.
+
+In this context, "matches" means _every path in the set, normalized through realpath and case sensitivity, begins with the path of at least one project folder in the Session under consideration_, and, as a special case, _a Session containing zero project folders matches everything_.
+
+The `--session` or `-s` flags to the `atom` command-line script explicitly designate the name of a Session which the newly opened window should open. If no Session with that name exists, a new one is created. If a window is already open on the named Session, the provided paths are added to the existing window and it is brought to the foreground.
+
+```sh
+# Open a new Atom window containing:
+# * the root folder "src/project-one"
+# * TextEditor on "${HOME}/writing/tps-report.txt"
+$ atom --session work src/project-one ~/writing/tps-report.txt
+
+# Open a new Atom window containing:
+# * the root folder "src/game"
+$ atom -s fun src/game
+
+# Add the root folder "src/project-three" to the existing "work" window
+$ atom -s work src/project-three
+
+# (close the "work" window)
+
+# Re-open a new Atom window with:
+# * root folders "src/project-one" and "src/project-three"
+# * TextEditor on "${HOME}/writing/tps-report.txt"
+$ atom -s work
+```
+
+`--session` is mutually exclusive with the `--add` option, which always adds the associated paths to the matching or last-opened window's Session.
+
+### Create a new bundled package to list and manage serialized state
+
+Clicking the "Session" tile in the tree-view opens a workspace item for listing, creating, loading, modifying, and deleting known Sessions. This is where Sessions can be renamed and managed.
+
+> TODO: graphic
+
+Loading a different Session immediately serializes the window's state into the previously active project and deserializes the window state from the chosen one into this window.
+
+A separate tab within the Session management shows known unsaved buffers. This acts as a kind of "dead letter office" to track down and recover lost changes.
+
+> TODO: graphic
+
+### When opening a buffer, restore its per-file state
+
+Each time a new buffer is opened, restore its per-file state if any has been persisted, regardless of the window's active session. If the file on disk has been modified externally since the unsaved changes were created, display a banner in the TextEditor to give the user a notice.
```

Comment:
Unfortunately, due to the way we serialize unsaved buffers, I don't think we have a good way to every recover these 😢 . I'd love to offer some kind of git-powered merge for the unsaved state.

Oh, unless there's something we can do with the files in `~/.atom/recovery/`?

Reply:
No idea how atom does it's serialization you talk about in respect to unsaved buffers, but why not use an actual git repo for that?
Then it would be easier to spot differences?